Collection of reverts used in 0.7.1  (REPLACED BY #457)

sukibaby opened the pull request:
## This is the complete collection of patches I reverted when making the 0.7.1 build. (not including version number or Simply Love rollbacks).


The 8/20/2024 test build and build 0.7.1 only differ in these reverted patches. Several testers remarked that the only sync issue remaining in 8/20/2024 is that the "center" of the sync is always shifting by ~6ms in either direction whereas 0.7.1 is fully stable.


### Inclusion of a patch here does not guarantee it affects sync negatively.  What IS guaranteed is that at least one of the patches (or some behavior due to several of these patches co-existing together) is responsible.

### I do not have the time to independently test all 15 of these, pull one out to retest and then distribute builds & collect data etc.  so I would prefer to remove any commits that can be _safely_ and _definitively_ ruled out from this batch of reverts.

sukibaby committed:
Revert "Remove RadianToDegree macro from ModelTypes"

This reverts commit 6a2de501011bf960bad932031637e7a43aa12094.

sukibaby committed:
Revert "Initialize SongPosition variables"

This reverts commit 74927934635513eac3ffea77465b395f6145df5c.

sukibaby committed:
Revert "WaveOut: Reattempt to initialize on failure"

This reverts commit b30bf853991f1128116a486dd78a57bfdafe6cc4.

sukibaby committed:
Revert "Prevent overlapping assist tick sounds"

This reverts commit e34450fbe81001e7b03692cee7e8d996a78cd0cd.

sukibaby committed:
Revert "Windows: Mark executable as DPI aware"

This reverts commit daad1a23337dcf1e7ed6e76880be171640c9f2ce.

sukibaby committed:
Revert "Don't necessarily hide the mouse cursor when full screened."

This reverts commit 481532ab7711046ba3a4e127ea453880f2786c77.

sukibaby committed:
Revert "Run as "High Priority" on Windows"

This reverts commit 3f8a7f5fd389cd3e8f923229bf46df278cdbf5cf.

sukibaby committed:
Revert "Windows: Set console output codepage to UTF-8"

This reverts commit cacaded0cd1804ed115fba6e1ce56e8cf3cec2c8.

sukibaby committed:
Revert "Windows: Remove old mapconv and verinc project files"

This reverts commit 3df79a3f11b097c662167db193eef38b6e0c1b74.

sukibaby committed:
Revert "Windows: Add manifest with UTF-8 code page specification"

This reverts commit 9a55cb99c1fd47618d05af9351ae6ad2e47a2366.

sukibaby committed:
Revert "Remove singlethreaded, old ffmpeg decoding code."

This reverts commit d74ad507079659e5b1edd7b3985da9257ba4d29e.

sukibaby committed:
Revert " Move LastOpenSection declaration outside of if/else, add clarity to comment"

This reverts commit bc65d16c4a1f1c07c427fc245f8cea23d8a4f4ab.

sukibaby committed:
Revert "Reset input redirection on screen change"

This reverts commit 17c4e7a2ff9869a09ee9b8928efef4a3ee029ea8.

sukibaby committed:
Revert "Fix coloring of BitmapText with overlapping attributes"

This reverts commit 208494b7358ccaa52f5b8d1efe1eabc632f55466.

sukibaby committed:
Revert "Applied time signature fix to SMALoader::ProcessBeatsPerMeasure()"

This reverts commit c4d63052e2389b593f58b13147523e3d114523db.

sukibaby commented:
## How Simply Love was ruled out

Testers were instructed to use Simply Love 5.5.0 or Zmod in 0.7.1 unless they had drift. Drift issues were not present which means the issue was not related to any change between Simply Love 5.3.0 and 5.5.0 (I believed, if there was an effect, it would be heavier resource use on the Lua backend causing issues, not the theme itself)

## How these patches were ruled out

The 8/11 test build is considered the starting point where the drift is finally almost eliminated for good. The difference in 8/11 and 8/20 test builds are the inclusion of #435 and #436 .

0.7.1 was released on 8/16. The only difference between 8/20 and 0.7.1 are the reverts featured in this PR.

We had multiple testers independently saying 8/20 feels stable except the "center" of the sync would randomly move around by ~6ms in either direction.

We had multiple testers saying 0.7.1 is not drifting.

mjvotaw commented:
I'm curious if you have any specific thoughts on https://github.com/itgmania/itgmania/pull/451/commits/c5826319f1d76b5e8587f681d94f3009dc4b1d03 and how it could contribute to drift issues.

A little background on that change:
I came across an issue while porting the tech counts/measure info/column cues from SL where a handful of step charts from ZiV had `#TIMESIGNATURES` tags that were missing an initial value for beat 0, like `#TIMESIGNATURES:196=2=4,198=4=4;`. Calling `TimingData::NoteRowToMeasureAndBeat()`for rows between 0 and the first defined time signature would crash. 
Since nowhere else in the engine that calls `TimingData::NoteRowToMeasureAndBeat()` seems to run into this problem, and it only affects songs with these sorts of "bad" time signature definitions, I figured it'd be easiest to just set the initial beat to 4:4 for them. For songs with no time signatures defined (which I'm pretty sure is most songs), this shouldn't change anything for them.

sukibaby commented:
> I'm curious if you have any specific thoughts on [c582631](https://github.com/itgmania/itgmania/commit/c5826319f1d76b5e8587f681d94f3009dc4b1d03) and how it could contribute to drift issues.


My reasoning for reverting c582631 is that it has the potential to indirectly affect when TimingData calls happen.

I very much don't suspect that patch but can't rule it out at this time. I also don't suspect patches like bc65d16, or 3df79a3, but I reverted them because there aren't many changes between 0.7.0 and 0.8.0 and they do have the potential to indirectly change behavior elsewhere.

In many ways I am going off the theory that the SM5 engine  is a big spaghetti mess of intertwined code so it's hard to rule out any particular part of the code as not effecting other places. Since I don't know some of these source files as well, I decided to play it safe by reverting anything which made a functional change that could potentially cause a change in behavior elsewhere. That's why I reverted the "Move LastOpenSection declaration" commit since it changes when a GAMESTATE call happens, even though I very much doubt `WheelBase.cpp` is responsible. Same goes for here why I reverted the `NoteLoaderSMA.cpp` change.

teejusb commented:
Fwiw I don't think I will go through and merge this because it reverts a bunch of things that I strongly do not think are relevant. I'd like more relevant statistics instead of anecdotes that these are actual regressions. Dimo for example says 0.9.0 (which is just the mainline release) has also been very consistent for him which already contradicts this PR so using a he-said she-said method isn't very helpful in actually determining root causes.

sukibaby commented:
> Fwiw I don't think I will go through and merge this because it reverts a bunch of things that I strongly do not think are relevant. I'd like more relevant statistics instead of anecdotes that these are actual regressions. Dimo for example says 0.9.0 (which is just the mainline release) has also been very consistent for him which already contradicts this PR so using a he-said she-said method isn't very helpful in actually determining root causes.

Did not mean for this to be merged as is, hoping to get a bunch removed from the list and only revert the relevant ones. The intention is to make a PR with only the ones to revert after discussing here

teejusb commented:
If you don't intend for this PR to be merged then mark it as a draft so I don't end up wasting time reviewing a PR that isn't ready to be merged.

sukibaby commented:
I want to get the discussion going because I have no further patches to propose. 

Dimo does not play test builds, so I'm not sure why you brought it up him liking 0.9.0 here @teejusb . We also have mac users saying they never experienced drift before 0.9.0, and the 0.7.1 patches fixed that issue on mac.

Remember recent tournaments that ran 0.9.0 on Windows experienced sync issues noted by multiple participants, but tournaments that ran 0.9.0 on Linux did not. On 0.8.0 Windows users were experiencing the most  issues but Linux users were not experiencing issues. SHINE is on 0.7.1 (so Windows), nobody experienced issues and several people got notable PB's.

mjvotaw commented:
Yeah, I think I'm down to go through these commits with you. I do think I generally agree with teejusb, in that it seems unlikely to me that most of these would be affecting the game's sync, but I'm open to being wrong about that.

Though one thing I'd like to do is go through some of the footage from these tournaments. I think, if there is something going on, a tournament would be the ideal place to catch it. 

I don't really know how tournaments work nowadays, is the footage from them generally available?

sukibaby commented:
> Yeah, I think I'm down to go through these commits with you. I do think I generally agree with teejusb, in that it seems unlikely to me that most of these would be affecting the game's sync, but I'm open to being wrong about that.

Thank you. I made this PR so we can discuss what may be worth reverting and what is definitely not a concern. What I had in mind when I made is PR is:

1. People can tell me what is listed here that definitely does not affect sync
2. People can tell me what is listed here they think doesn't affect sync

I'll make a comment after this one focusing on that so it's more organized.

> Though one thing I'd like to do is go through some of the footage from these tournaments. I think, if there is something going on, a tournament would be the ideal place to catch it.
> 
> I don't really know how tournaments work nowadays, is the footage from them generally available?

Yes, all the VODs from SHINE are on twitch.tv/intltiming and also I intend to download them in case the VODs disappear in the distant future. Drift was never once a concern thru the tournament with the in-game timer showing upwards of 8 hours, no issues or desync.

I am not sure where the VODs from BITE are but that was the tournament that ran 0.9.0 on Windows and it's documented that there were sync drift concerns during the event.

My understanding that 0.9.0 on Linux was considered fine/drift free in a tournament setting was when @teejusb  mentioned that the **Project Storm** machines were running 0.9.0 on Ubuntu.

teejusb commented:
> I want to get the discussion going because I have no further patches to propose.
> 
> Dimo does not play test builds, so I'm not sure why you brought it up him liking 0.9.0 here @teejusb .
> 
> Remember recent tournaments that ran 0.9.0 on Windows experienced sync issues noted by multiple participants, but tournaments that ran 0.9.0 on Linux did not. On 0.8.0 Windows users were experiencing the most issues but Linux users were not experiencing issues. SHINE is on 0.7.1 (so Windows), nobody experienced issues and several people got notable PB's.

I brought it up because dimo was one of the original people having significant issues with drift (and one of the original posters of the thread). It's also relevant because a lot of commits being proposed to be reverted in this PR are included in 0.9.0. I'm not against reverting known problematic commits, but being more stringent about *which* commits instead of doing broad strokes I think is important. Some of the proposed reverts def seem like a step backwards.

teejusb commented:
Yeah, Storm was on 0.9.0 and no mention of drift, but it was also on Linux fwiw.

sukibaby commented:
@teejusb 
@mjvotaw 

### First of all, please let me re-establish that the only issue within one of the patches here is that it moves the center of the sync off by a few ms every song or every other song. Within the song people report it is stable. People have been compensating for this on the affected builds by adjusting their visual delay.

There is NO other sync related concern in the following patches.

## My patches I reverted and why
1. Revert "Remove RadianToDegree macro from ModelTypes" 224f829
	- Affects a function "LoadMilkshapeAsciiBones" which is only called by
	    - BeginnerHelper.cpp
		- DancingCharacters.cpp
		- Model.cpp / .h
		- ModelTypes.cpp
		- ScreenHowToPlay.cpp
	- Is it affecting sync? I doubt it, but maybe it's affecting how fast Model.cpp/.h runs.

2. Revert "Initialize SongPosition variables" d32606e
	- Is it affecting sync? I doubt it, unless initializing all the values to 0 is wasting time and causing SongPosition to respond slower.

3. Revert "WaveOut: Reattempt to initialize on failure" abad7ff
	- Is it affecting sync? It shouldn't but adds the tiniest bit of delay to the WaveOut driver so it's possible?

## Patches I reverted this more because i was rolling back all non essential changes than I suspected it was affecting sync
1. Revert "Prevent overlapping assist tick sounds" dbfb713
	- Is it affecting sync? No way that I can imagine, since it shouldn't make any difference unless assist tick is on. I reverted it because I was rolling back all non essential changes.

2. Revert "Don't necessarily hide the mouse cursor when full screened." 3c3727e

3. Revert "Run as "High Priority" on Windows" d848835

4. Revert "Remove singlethreaded, old ffmpeg decoding code." a31fb7f

5. Revert " Move LastOpenSection declaration outside of if/else, add cla… 9b9f07e

6. Revert "Applied time signature fix to SMALoader::ProcessBeatsPerMeasu… c582631

7. Revert " Move LastOpenSection declaration outside of if/else, add cla… 9b9f07e


## Patches which I don't understand how the following changes affect the engine on a deeper level, but we did find that the "Compile as UTF-8" flag that was used here was causing a major dip in performance, and it was reverted, so I am a little suspicious of these other commits from the same PR.
1. Revert "Windows: Mark executable as DPI aware" 69f04fe

2. Revert "Windows: Set console output codepage to UTF-8" 1bc20e6

3. Revert "Windows: Remove old mapconv and verinc project files" 12e127e

4. Revert "Windows: Add manifest with UTF-8 code page specification" c12b7cf


## Patches I suspect may be affecting sync but don't have proof
1. Revert "Reset input redirection on screen change" 4d28b5b
	- Changes some fairly low level behavior. Smaller changes have caused big problems. Don't know how well this was tested to make sure the engine works the same with this patch.

2. Revert "Fix coloring of BitmapText with overlapping attributes" 96e9ee1
	- Changes the behavior BitmapText::AddAttribute, a fairly low level and commonly called function

sukibaby commented:
If I had to guess  then the ones I suspect the most would be 

- Reset input redirection on screen change
- Initialize SongPosition variables

sukibaby commented:
After further consideration I do actually consider **Revert "Applied time signature fix to SMALoader::ProcessBeatsPerMeasu..."**  https://github.com/itgmania/itgmania/commit/c5826319f1d76b5e8587f681d94f3009dc4b1d03 to be the relevant patch.

1. It changes `ProcessBeatsPerMeasure` to store segments in a vector and add all segments to `TimingData` after all have been processed. The original function adds each `TimeSignatureSegment` to `TimingData` immediately after creation.
2. If the first segment does not start at row 0, a default time signature (4/4) is added at the beginning.
3. It changes behavior of `AddSegment` which may result in different results than the method it replaced

The implications are visual offsets due to inconsistencies in how timing data is processed.

Additional evidence:

1. We know the relevant patch was most likely committed to 0.8.0-release due to complaints about sync increase starting with 0.8.0.
2. The issue of sync shifting slightly **between each song** would make sense for this.
3. Testers said this problem was solved by adjusting the visual offset slightly after each song, and changes in this function's behavior directly affect the visual offset of the song displayed.

teejusb commented:
One major thing and I'm curious about is that code in that commit is only run on `.sma` files. I don't know about you but I'm pretty sure I've *never* even seen an `.sma` file -- it's usually `.sm`/`.ssc`, and a couple of `.dwi`. My machine (which I suppose has /only/ a lowly 15,000 songs) has none.

![image](https://github.com/user-attachments/assets/e75f4553-a8b2-453b-9afa-710cbe769ca7)

sukibaby commented:
> One major thing and I'm curious about is that code in that commit is only run on `.sma` files. I don't know about you but I'm pretty sure I've _never_ even seen an `.sma` file -- it's usually `.sm`/`.ssc`, and a couple of `.dwi`. My machine (which I suppose has /only/ a lowly 15,000 songs) has none.
> 
> ![image](https://private-user-images.githubusercontent.com/5017202/361647925-e75f4553-a8b2-453b-9afa-710cbe769ca7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3MzIzODMsIm5iZiI6MTcyNDczMjA4MywicGF0aCI6Ii81MDE3MjAyLzM2MTY0NzkyNS1lNzVmNDU1My1hOGIyLTQ1M2ItOWFmYS03MTBjYmU3NjljYTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjdUMDQxNDQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk3MzIwZjk5MTMzZjhjOTkzNGZjZmFmN2I2ZGUyMmEwNDI0YTAwNmFhMjU0YzFlMzMxNmEyMGZiZjVhMmE0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.a-GFoPDMUjLplV6BBDNYxIyxyjYj0yN2_VlXL75xiQc)

Hmmmmmmm...  I was fundamentally confused about the purpose of `NotesLoaderSMA` as I've never seen or heard of a SMA file ever, so I thought it was used in conjunction with `NotesLoaderSM` since they had non-overlapping sets of functions. So I mistakenly tested with a sm file.

So you're right that it can't be this.

~~It also wouldn't affect ssc's which all of ITL 2024 was so it can't be this.~~

sukibaby commented:
Wait, sorry,  I confused myself. There was a [similar change made to NotesLoaderSM that I should have patched instead](https://github.com/itgmania/itgmania/commit/eeb25286127cb6acaa6cc6b74cb1c46722870355) ~~but that still doesn't explain sync issues experienced on SSC files~~.

edit: Actually I added some logging to `SMLoader::ProcessTimeSignatures` to let me know when it processed a file and indeed SSC's are processed by this, as NotesLoaderSM is included by NotesLoaderSSC.

teejusb commented:
iirc `.sma` are used it for some PIU style charts, which ITG wouldn't fall under

sukibaby commented:
> iirc `.sma` are used it for some PIU style charts, which ITG wouldn't fall under

Thanks, good to know.

So I still can't imagine what else from 0.8.0 could have ANY impact on sync: https://github.com/itgmania/itgmania/compare/v0.7.0...v0.8.0

I didn't revert borderless fullscreen mode here, so it's not that, it's probably not any of the CourseWriter related changes as those aren't used by the majority of players, and @CrashCringle12 's changes seem to affect the songwheel exclusively, so why did Windows users start noticing sync issues on 0.8.0?

sukibaby commented:
NotesLoaderSSC does pull in NotesLoaderSM so is there some chance the commit which affects NotesLoaderSM is responsible? I can't imagine what else there could impact visual delay.

sukibaby commented:
I think I can rule out ["Initialize SongPosition variables"](https://github.com/itgmania/itgmania/pull/451/commits/d32606edf3b549afe03461b9f390fb263f2446bf), because adding logging to that change indicates that it's executed a few times during game initialization and doesn't seem to be called again.

I've ruled out ["Remove RadianToDegree macro from ModelTypes"](https://github.com/itgmania/itgmania/pull/451/commits/224f829dcf567e701780d754da56fc3144f3064a) as logging confirms it's never called during typical gameplay in Simply Love.

I've ruled out ["Fix coloring of BitmapText with overlapping attributes"](https://github.com/itgmania/itgmania/pull/451/commits/96e9ee1bcd73f0ca1f6a20545176d587e783ad0f) for the same reason. Seems to only be called once early on after the songs are loaded and not after that.

I'm ruling out ["WaveOut: Reattempt to initialize on failure"](https://github.com/itgmania/itgmania/pull/451/commits/abad7ff0e7dc0a8abd56b34ca00faac1234febb9) because the only changes happen within a conditional check that's being skipped unless an audio driver failure worthy of crashing the game happens, but i did profile it with a timer just to be safe.

I'm ruling out ["Prevent overlapping assist tick sounds"](https://github.com/itgmania/itgmania/pull/451/commits/dbfb7137f7d54b5d128744daabcf4acd1363a3ad) because it doesn't change anything when assist clap isn't on.

I don't suspect ["Don't necessarily hide the mouse cursor when full screened."](https://github.com/itgmania/itgmania/pull/451/commits/3c3727e1c9598e78aa20a8af0a7d3f1dc7b3a8bb) though it's possible it might change some unknown behavior on the Windows side of things.

I don't suspect the FFMpeg code because that was only reverted to go back to 0.7.0 behavior.

I'm ruling out ["Applied time signature fix to SMALoader::ProcessBeatsPerMeasu…](https://github.com/itgmania/itgmania/pull/451/commits/c5826319f1d76b5e8587f681d94f3009dc4b1d03) because, even though it's the likely candidate, it seems like it's not affecting anything it shouldn't when adding logging to verify what it's affecting.

## So this is my revised list of patches I suspect...

**Revert "Windows: Mark executable as DPI aware"**
   -  suspected because it changes low level behavior between the game and the OS.
   -  i objected to this change when the PR was made because the game is not actually aware of the DPI so we can't know exactly what unseen changes this made.

**Revert "Run as "High Priority" on Windows"** 
   -  suspected because it might be leading to an incorrect thread priority structure


**Revert "Windows: Set console output codepage to UTF-8" 
Revert "Windows: Remove old mapconv and verinc project files" 
Revert "Windows: Add manifest with UTF-8 code page specification"** 
   -  suspected because they were from the same PR as "MSVC: Compile as UTF-8"


**Revert "Reset input redirection on screen change"** 
   -   suspected for changing low level screen behavior

**Revert Move LastOpenSection declaration outside of if/else, add cla…**
   -   moves a GAMESTATE call, has the potential to be a regression surrounding state changes

sukibaby commented:
I've ruled out ["Remove RadianToDegree macro from ModelTypes"](https://github.com/itgmania/itgmania/pull/451/commits/224f829dcf567e701780d754da56fc3144f3064a) as logging confirms it's never called during typical gameplay in Simply Love.

sukibaby commented:
We can continue the conversation on https://github.com/itgmania/itgmania/pull/457 .
